### PR TITLE
Assignment 1 PR for Feedback

### PR DESCRIPTION
**Pull request by gicamikiel**
UPDATE: Reimplementation completed, commits linked below
https://github.com/gicamikiel/Computing-at-Scale/commit/d7c9b3837a2a8f3f9e91bb7566c9f663f526abec
https://github.com/gicamikiel/Computing-at-Scale/commit/8250d05820a0bd4ba4ce5967b921278be0675651
(no one brought this up but this was bugging me)
https://github.com/gicamikiel/Computing-at-Scale/commit/c10b254b8d661e2d341475374498473fd099374b

### CODE REVIEW

File: assignment1/README.md
Comment (ickbumk) [inline assignment1/README.md:1]: Everything compiled nicely, and the instructions were clear. The unittest works nicely. Chebyshev fails for all cases.

File: assignment1/source/integrationrule.cpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    return (to-from)/2.0;
+}
+
+double GaussLegendre::xi_to_x(double xi, double from, double to) const {
+    return (to-from)*xi/2.0 + (from+to)/2.0;
+}
+
+double GaussLobatto::operator()(ScalarFunction& function, double from, double to) const {
+    double out = 0;
+    for (int i=1; i<num_pts()-1; i++) { // min number of num_pts is 3
```
Comment (ickbumk) [inline assignment1/source/integrationrule.cpp]: std::size_t (I think) would better suit for indexing

File: assignment1/source/integrationrule.cpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
+#include <stdexcept>
+#include <cmath>
+
+#include "integrationrule.hpp"
+
+double IntegrationRule::operator()(ScalarFunction& function, double from, double to) const {return 0;}
+
+GaussLegendre::GaussLegendre() : xi_pts_({}), weights_({}) {}
```
Comment (ickbumk) [inline assignment1/source/integrationrule.cpp]: I think the constructor does not need empty {} since the default is an empty state?

File: assignment1/source/integrationrules.hpp [hunk 3]
```diff
@@ -0,0 +1,30 @@
+#include <map>
+#include <string>
+#include <cmath>
+
+#include "integrationrule.hpp"
+
+// File for defining common quadrature rules
+
+#define HI_WT_PT std::sqrt(3.0/7 - 2.0/7*std::sqrt(6.0/5))
```
Comment (ickbumk) [inline assignment1/source/integrationrules.hpp:9]: we can try "constexpr" instead of #define

File: assignment1/source/polynomial_integrate.cpp [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        }
+        rule = &glo_rules[num_pts];
+    } else if (rulename == "gauss-chebyshev") {
+        rule = &gc;
+    } else {
+        std::cout << "Invalid rule name\n";
+    }
+
+    double out = (*rule)(p, from, to);
+    std::cout << out << "\n";
```
Comment (ickbumk) [inline assignment1/source/polynomial_integrate.cpp:70]: return 0 maybe?

**Review by ickbumk — COMMENTED** [review]
Does the style match?
Yes

Is there documentation of the new code? Does it address what it does and how to use it?

I think the code will definitely improve with more module docstring and function docstrings. I was able to follow the code without significant difficulties, just because I did the same assignment. However, if someone who is not familiar with the assignment may have hard time understanding what's going on.

Are there any obvious bugs, or performance issues?

No bugs or performance issues.

Is the code easy to read and understand? Is it well organized? Are function names clear?

all function names were clear. No abbreviations were used. The header files are well organized .

Can raw loops be replaced with standard library algorithms?

Nothing that I am aware of

General questions or concerns.

No concerns. I am impressed how the unittest was constructed. My implementation is far primitive than this. 


Does the code compile?

Yes

Are there any compiler warnings?

No

Are there any runtime errors?

No

Does the code do what it is supposed to do?

Yes. The unittest also performs test for all possible cases. 

Are there test cases for all new functionality?

Unittest thoroughly tests for all cases.

Are there test cases that you think should be added?

No

Do all of the tests pass?

No. Chebyshev does not pass the test. 

Are there any performance regressions?

Not that I am aware of

Does the code scale to the expected problem size?

We have not tried this problem with more than 1 dimensional case or polynomials with very high degree. This is not applicable.

Great code overall. Impressive.

**Review by scottblender — COMMENTED** [review]
- Does the style match the rest of the codebase? (naming conventions, indentation, etc.) Tools like clang-format can help.
    - Yes, the style is consistent.
- Is there documentation of the new code? Does it address what it does and how to use it?
    - Yes, the documentation is fine. The README.md could include much more information, though, about how each base class and derived class works, how to input information into how to use polynomial_integrate, and what methods/functions are called.
    - Code can also have more comments to explain functionality, even if it should be easy to read/understand.
- Are there any obvious bugs or performance issues?
    - No, there are no apparent bugs or performance issues.
- Is the code easy to read and understand? Is it well-organized? Are function names clear?
    - Yes, the code is easy to read and well-organized.
- Can raw loops be replaced with standard library algorithms?
    - Mostly no. It might be worthwile to look at `std::for_each` on cppreference.
- Does the code compile?
    - Yes.
- Are there any compiler warnings?
    - No.
- Are there any runtime errors?
    - No.
- Does the code do what it is supposed to do?
    - Yes.
- Are there test cases for all new functionality?
    - Yes.
- Are there test cases that you think should be added? i.e., corner cases that are not covered.
    - Gauss-Legendre and Gauss-Lobatto work well. For Gauss-Chebyshev, since the weighting function is sqrt(1-x^2) or 1/sqrt(1-x^2), I would modify this to only work to integrate functions of this type in the specified bounds. I believe the change of coordinates for this quadrature method is a bit more complex since h(x) = f(x)/sqrt(1-x^2) and dh/dxi becomes slightly more complicated when scaling the integral. 
    - For the specific test cases, I like that you tested all polynomials the quadrature method can exactly integrate.  In your specific quadrature methods, one thing to think about is throwing a `std::invalid_argument` if the incorrect number of points are sent in or if the polynomial is out of bounds.
- Do all of the tests pass?
    - No. Chebyshev is incorrect, but this can be solved by simplfying it to only work for -1 to 1.
- Are there any performance regressions?
    - No, performance is fine. To improve performance, it might be better to test edge cases rather than every single possible polynomial. 
- Does the code scale to the expected problem size?
    - Yes, the code is designed to scale well.

File: assignment1/README.md [hunk 5]
```diff
@@ -0,0 +1,9 @@
+# Assignment 1
+## Build Instructions
+1. Navigate into ```assignment1``` directory
+2. Run ```cmake -S . -B build``` in the terminal
+3. Run ```cmake --build build``` to compile
+
+## Usage
+* Run ```build/source/polynomial_integrate``` to get usage information for integrating polynomials in the command line.
```
Comment (scottblender) [inline assignment1/README.md:8]: What are the inputs and input orders for this? Could you provide instructions in the README.md in case someone trying to use this could not see the source code?

File: assignment1/source/integrationrule.cpp (hunk 1 above)
Comment (scottblender) [reply assignment1/source/integrationrule.cpp]: you could also use auto.

File: assignment1/source/integrationrule.cpp [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        out += weights_[i]*function(xi_to_x(xi_pts_[i], from, to));
+    }
+    double lbound = function(xi_to_x(-1.0, from, to));
+    double rbound = function(xi_to_x(1.0, from, to));
+    double bounds = 2.0/num_pts()/(num_pts()-1)*(lbound+rbound);
+    return dx_dxi(from, to)*(bounds+out);
+}
+
+GaussChebyshev::GaussChebyshev(int num_pts) {
+    if(num_pts <= 0) throw std::invalid_argument("Error: cannot have negative or zero number of points");
```
Comment (scottblender) [inline assignment1/source/integrationrule.cpp]: I would add in cases where you enter more than 4 points too.,

File: assignment1/source/integrationrules.hpp
Comment (scottblender) [inline assignment1/source/integrationrules.hpp:1]: I would rename this to something a little more clear. It might be confusing for someone looking at both the base class and the derived class.

File: assignment1/source/integrationrules.hpp [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+#define HI_WT (18+std::sqrt(30))/36.0
+#define LOW_WT_PT std::sqrt(3.0/7 + 2.0/7*std::sqrt(6.0/5))
+#define LOW_WT (18-std::sqrt(30))/36.0
+
+std::unordered_map<int, GaussLegendre> gl_rules = {
```
Comment (scottblender) [inline assignment1/source/integrationrules.hpp]: nice use of `std::unordered_map`

File: assignment1/source/scalarfunction.cpp [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    double sum = 0;
+    for (const auto [exp, coef] : coefs_) {
+        //std::cout << coef << "*" << x << "^ " << exp << " the sum is " << sum << "\n";
+        sum += coef*std::pow(x,exp);
+    }
+    return sum;
+}
+
+double& PolynomialFunction::operator[](int exp)  {
+    if (exp < 0) throw std::invalid_argument("Error: polynomials cannot have negative exponents");
```
Comment (scottblender) [inline assignment1/source/scalarfunction.cpp:33]: Nice check for negative exponents. What about fractional exponents, too?

File: assignment1/source/integrationrule.cpp
Comment (scottblender) [inline assignment1/source/integrationrule.cpp:1]: Using templates for the return type of the () operators or wherever you initialize a std::vector might be worthwhile.

File: assignment1/source/integrationrule.hpp [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        GaussLegendre();
+
+        GaussLegendre(std::vector<double> xi_pts, std::vector<double> weights);
+
+        double operator()(ScalarFunction& function, double from, double to) const override;
+
+        int num_pts() const;
+
+    protected: 
+        std::vector<double> const xi_pts_;
```
Comment (scottblender) [inline assignment1/source/integrationrule.hpp:24]: Should these be const? Might the user want to change the values of these after initializing them?

File: assignment1/source/integrationrules.hpp [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    {1, GaussLegendre({0}, {2.0})},
+    {2, GaussLegendre({-1.0/std::sqrt(3), 1.0/std::sqrt(3)}, {1.0, 1.0})},
+    {3, GaussLegendre({-std::sqrt(3.0/5), 0, std::sqrt(3.0/5)}, {5.0/9, 8.0/9, 5.0/9})},
+    {4, GaussLegendre({-LOW_WT_PT, -HI_WT_PT, HI_WT_PT, LOW_WT_PT}, {LOW_WT, HI_WT, HI_WT, LOW_WT})}
+};
+
+std::map<int, GaussLobatto> glo_rules = {
```
Comment (scottblender) [inline assignment1/source/integrationrules.hpp]: Why use a regular `std::map` compared to a `std::undordered_map` for Gauss-Lobatto?

File: assignment1/source/polynomial_integrate.cpp [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+    IntegrationRule* rule;
+    if(num_pts < 1) {
+        std::cout << "At least one point needed\n";
+        return 1;
+    }
+    GaussChebyshev gc(num_pts);
+    if(rulename == "gauss-legendre") {
+        if(num_pts > 4) {
+            std::cout << "Only 1-4 points are supported\n";
```
Comment (scottblender) [inline assignment1/source/polynomial_integrate.cpp:53]: I think implementing this in the class itself is "safer."

File: assignment1/source/integrationrules.hpp (hunk 3 above)
Comment (gicamikiel) [reply assignment1/source/integrationrules.hpp:9]: I used macros here because I did not want these to be used outside of the file.

Macros are fairly ugly though and I realized I might be able to limit this variable to one file using an anonymous namespace? I will try a constexpr variable in one.

File: assignment1/source/scalarfunction.cpp (hunk 8 above)
Comment (gicamikiel) [reply assignment1/source/scalarfunction.cpp:33]: That is a good idea, I did not think about fractional exponents

File: assignment1/source/integrationrules.hpp (hunk 10 above)
Comment (gicamikiel) [reply assignment1/source/integrationrules.hpp]: They should be the same, I think I made a mistake here

File: assignment1/README.md (hunk 5 above)
Comment (gicamikiel) [reply assignment1/README.md:8]: I relied on writing and printing a good help message when the integrate polynomial executable was run, which means you technically do not need to look at the source code. 

When people use new software on Linux they usually check the help message when you run the executable, but you are right that I should add instructions in the README as it would be better.

File: assignment1/source/polynomial_integrate.cpp (hunk 11 above)
Comment (gicamikiel) [reply assignment1/source/polynomial_integrate.cpp:53]: I kept the definition for the maximum number of points outside of any classes because a user of the library can technically define more integration points for their own purposes.

File: assignment1/source/integrationrule.hpp (hunk 9 above)
Comment (gicamikiel) [reply assignment1/source/integrationrule.hpp:24]: The intent I was going for here was that if someone needs the values changes they can create another instance with different values. I don't expect lots of IntegrationRule classes to be created on the fly so this should not be an issue. Maybe I could have made this clearer.

File: assignment1/source/integrationrule.cpp (hunk 2 above)
Comment (gicamikiel) [reply assignment1/source/integrationrule.cpp]: It seems like a get a compile error if I leave out the empty {}

File: assignment1/source/scalarfunction.cpp (hunk 8 above)
Comment (gicamikiel) [reply assignment1/source/scalarfunction.cpp:33]: Actually I just realized passing in a fractional exponent should not be possible because the parameter data type is an integer
